### ISSUE TITLE
Peer discovery via Weave Cloud

### ISSUE THREAD

**Issue by brb** [weaveworks/weave#2799]
As a user I want to give each peer a token to simplify cluster configuration, for instance `weave launch --token ABCDEABCDEABCDE`.

A peer started in such way would connect to Weave Cloud to retrieve list of peers to which it has to connect.

Before starting implementing this feature we need to:

1. Refine requirements.
2. Create a design doc.

**Comment by bboreham** [comment]
Requirements:

## Goal
Currently, all users of Weave Net are required to supply each peer with a list of some other peers.  This task gets complicated when you consider adding and removing machines, and there is a fair chance for things to go wrong.

Weave Cloud Peer Discovery will offer users an alternative: they obtain from Weave Cloud a token, give that token to each peer in the cluster, and those peers will thereby connect to each other.

## User Stories
### Day One
* Obtaining a token for user’s first cluster
* Obtaining a token for user’s nth cluster
* Configuring Weave Net given a token
* Peers that can see other peers directly form a Weave Net cluster
* Day One point one
* Removing a dead peer from a cluster, via the CLI
* View a simple list of peers in GUI
* Peers that are behind a NAT gateway get joined to an existing Weave Net cluster

### Day Two
* Move a peer from one cluster to another
* View an audit log of the cluster (peers joining, leaving, etc.)
* View "dead" peers in GUI; use button to remove from list
* Control use of public vs private addresses
* Gui to configure parameters e.g. IP allocation range
* Hybrid clusters, e.g. peer discovery through Docker Swarm or kube-peers and Weave Cloud
* See peers known to peer discovery via Topology View in Weave Cloud

### Day Three
* Weave Cloud Peer Discovery removes gone-away peers without user interaction
* Weave Cloud can form an efficient topology for hundreds to thousands of peers
* Receive notification of a peer joining or leaving the cluster (e.g. a webhook)

## Constraints and requirements
### Must
* Provide logs to operators for troubleshooting
* Support day one number of peers per cluster <100; total across all clusters 10,000
* QPS: purely to illustrate, suppose 7,200 peers started in a 5-minute period; that’s 20qps (across all clusters)
* Add no more than 2 seconds to start-up time for 99% of peer starts
* Handle two peers joining at the same time (implies consistent data store)
* Not blow up when connectivity from peers is intermittent
* Not blow up when connectivity between peers is intermittent
* Not blow up when a peer is shut down for a time then restarted
### Nice to have
* Be able to wind back data store to a previous snapshot in case of bugs/fat-finger
* Scaling to one million peers (across all clusters, i.e. a few with 100-1000 peers, more with 10-100, and lots with 1-10)
* Tighter latency bounds
### Out of scope
* Weave-DNS (does not seem to be impacted by cloud)

(copied from https://docs.google.com/document/d/18I47KYnfcGHgUiQSXdPJ2NIJaY10uDvK6HYQxSbbWU0/ which is not publicly viewable)
